src/search_stream.rs

```diff
@@ -203,6 +213,9 @@ impl<'a, R: io::Read, W: Terminal + Send> Searcher<'a, R, W> {
                 break;
             }
             while self.inp.pos < self.inp.lastnl {
+                if self.opts.files_with_matches && self.match_count > 0 {
+                    break;
```

Comment:
Oh, yes, indeed you do. The outer loop corresponds to refilling the buffer with data from the file and the inner loop corresponds to iterating over all matches within that buffer. Breaking out of the inner loop won't cause the outer loop to stop.

You might consider changing `loop` to `while !self.opts.files_with_matches || self.match_count == 0` or something similar. (With the case unfortunately repeated inside the `while` loop as well.)